test/e2e/network_policy.go

```diff
@@ -0,0 +1,438 @@
+	// for pod discovery by the net monitor containers.
+	serviceA := createNetworkPolicyService(f, nsA, serviceAName)
+	serviceB := createNetworkPolicyService(f, nsB, serviceBName)
+
+	// Clean up services
+	defer func() {
+		By("Cleaning up the service A")
+		if err = f.Client.Services(nsA.Name).Delete(serviceA.Name); err != nil {
+			framework.Failf("unable to delete svc %v: %v", serviceA.Name, err)
+		}
+	}()
+	defer func() {
+		By("Cleaning up the service B")
+		if err = f.Client.Services(nsB.Name).Delete(serviceB.Name); err != nil {
+			framework.Failf("unable to delete svc %v: %v", serviceB.Name, err)
+		}
+	}()
+
+	By("Creating a webserver (pending) pod on each node")
+	podA, podBs := launchNetMonitorPods(f, nsA, nsB, nodes)
+
+	// Deferred clean up of the pods.
+	defer func() {
+		By("Cleaning up the webserver pods")
+		if err = f.Client.Pods(nsA.Name).Delete(podA.Name, nil); err != nil {
+			framework.Logf("Failed to delete pod %v: %v", podA.Name, err)
+		}
+		for _, podB := range podBs {
+			if err = f.Client.Pods(nsB.Name).Delete(podB.Name, nil); err != nil {
+				framework.Logf("Failed to delete pod %v: %v", podB.Name, err)
+			}
+		}
+	}()
+
+	// Wait for all pods to be running.
+	By(fmt.Sprintf("Waiting for pod %v to be running", podA.Name))
+	err = framework.WaitForPodRunningInNamespace(f.Client, podA)
+	Expect(err).NotTo(HaveOccurred())
+	for _, podB := range podBs {
+		By(fmt.Sprintf("Waiting for pod %v to be running", podB.Name))
+		err = framework.WaitForPodRunningInNamespace(f.Client, podB)
+		Expect(err).NotTo(HaveOccurred())
+	}
+
+	// Open up TCP port used to access the network monitor HTTP API (8080).  Until we do this, all
+	// pods in all namespaces are isolated and the test cannot query the netmonitor pods to pull
+	// the connectivity information.
+	By("Open up monitoring port so that test can pull connectivity data from services")
+	addNetworkPolicyOpenPort(f, nsA, "query1", 8080, api.ProtocolTCP)
+	addNetworkPolicyOpenPort(f, nsB, "query2", 8080, api.ProtocolTCP)
+
+	// We expect full isolation between all pods in all namespaces.  Monitor, wait for a little and
+	// re-check we are still isolated.
+	By("Checking full isolation")
+	expected := Summary{
+		TCPNumOutboundConnected: 0,
+		TCPNumInboundConnected:  0,
+	}
+	monitorConnectivity(f, nsA, serviceA.Name, expected)
+	monitorConnectivity(f, nsB, serviceB.Name, expected)
+
+	By("Waiting and rechecking full isolation")
+	time.Sleep(10 * time.Second)
+	monitorConnectivity(f, nsA, serviceA.Name, expected)
+	monitorConnectivity(f, nsB, serviceB.Name, expected)
+
+	if nsA != nsB {
+		// Add policy to one namespace to accept all traffic to the TCP port used for inter-pod pings (8081).
+		// We should see mono-directional traffic.  We only do this if the two namespaces are different - since
+		// we open up ports on a per-namespace basis.  If the namespaces are identical we'll end up straight
+		// away with bi-directional traffic.
```

Comment:
You should be able to do a bi-directional test when the two services are in the same namespace by specifying a pod selector which selects each service (rather than selecting an entire namespace).

This means the policies you create for the single namespace vs multi-namespace case will be different, but buys you actually testing `podSelector` policies.